### ISSUE TITLE
MultiAppTransfer "execute_on" might need re-design

### ISSUE THREAD

**Issue by aeslaughter** [idaholab/moose#4649]
Currently, `MultiAppTransfer` adds a special "SAME_AS_MULTIAPP" option to "execute_on". This option is then stripped from the "execut_on" parameter before calling the `Transfer` base class constructor. This will be a problem for #4604 and #4606 because the stand-along function that performs the strip of the option requires a const_cast to modify the parameter.

I propose:
- Adding a new parameter: "used_multiapp_execute_on" that defaults to true.
- If this new option is true, then utilize "execute_on" from the `MultiApp`.
- If it is true AND "execute_on" is set by the user produce an error
- If it is false, use the "execute_on" from the `Transfer` and produce an warning that the execution times are different between the `MultiApp` and the `MultiAppTransfer`, if they are different.

**Comment by friedmud** [comment]
Are you proposing adding this to ALL objects?

**Comment by permcody** [comment]
No, he doesn't like the const cast in Transfers. It's low priority if you
ask me since it's just another solution that's not really more clear but
does fix the odd code path necessary to pull this off. Since users don't
see the code I personally think it's fine.

**Comment by friedmud** [comment]
I just don't understand how this changes anything.  He says:

> Adding a new parameter: "used_multiapp_execute_on" that defaults to true.

Adding it where?  That's my question.  Is it adding it to the basic set of `execute_on` parameters?  We're already adding `SAME_AS_MULTIAPP`... so how is this different?

**Comment by permcody** [comment]
He wants another parameter that you'd set instead of execute_on

**Comment by friedmud** [comment]
Then I really vote against that.  The users I know that are using this new functionality (having `execute_on` set to `SAME_AS_MULTIAPP`) are really enjoying it... and think it's really intuitive.  It was a HUGE upgrade over what was there before.

I don't have any problems with a const_cast in this case either... although we need to think about how this will be handled in parallel (are all of the thread parallel objects going to be referencing the same InputParameters reference?  If so... then we'll need to be careful here... but that's a discussion that needs to be in #4606)

**Comment by aeslaughter** [comment]
I will keep the const_cast.
